Apply general feedback for `@sources`

RobinMalfait opened the pull request:
Part of #14078

This PR applies general feedback to the `@sources` umbrella PR.

This PR includes the following changes:
- Used real classes instead of class-like objects.
- Only handle PostCSS imports when `@import` is used.
- Improve some comments
- Improve naming of `GlobEntry` (`glob.glob` -> `glob.pattern`)
- Use `scanFiles` on the `scanDirResult` as-if it's an incremental scan (under
  the hood we still do a full scan right now).
- Renamed `content_paths` to `sources`

This is a separate PR just so that we can merge the umbrella PR un-squashed for each sub-PR. This way all the commits here won't show up in the `next` branch as-is.

RobinMalfait committed:
make `disposables` a `Disposables` class

RobinMalfait committed:
use `let`

RobinMalfait committed:
only check for `@import`

We don't need to fix relative paths if we are not importing. We only
care about `@import`.

RobinMalfait committed:
use correct import for `internal-postcss-fix-relative-paths`

RobinMalfait committed:
improve comment

RobinMalfait committed:
use more generic `callback` name

RobinMalfait committed:
rename `content_paths` to `sources`

RobinMalfait committed:
use `scanDirResult.scanFiles()`

Internally in Rust, we still use the `scanDir()` function, and the
`scanFiles()` function will be optimized later but the clients can use
the optimal API.

RobinMalfait committed:
use correct `root` path

RobinMalfait committed:
rename `glob` in `GlobEntry` to `pattern`

RobinMalfait commented on crates/node/src/lib.rs:
```diff
@@ -32,35 +38,37 @@ pub struct ScanResult {
-      input.into_iter().map(Into::into).collect(),
-      self.globs.clone().into_iter().map(Into::into).collect(),
-    )
+  pub fn scan_files(&self, _input: Vec<ChangedContent>) -> Vec<String> {
+    let result = tailwindcss_oxide::scan_dir(tailwindcss_oxide::ScanOptions {
+      base: self.base.clone(),
+      sources: self.sources.clone().into_iter().map(Into::into).collect(),
+    });
+
+    result.candidates
   }
```
Public API uses `scanDirResult.scanFiles()` for incremental rebuilds. Right now we will do a full scan behind the scenes.

Good idea, @philipp-spiess!

RobinMalfait commented on crates/node/src/lib.rs:
```diff
@@ -24,6 +24,12 @@ impl From<ChangedContent> for tailwindcss_oxide::ChangedContent {
 #[derive(Debug, Clone)]
 #[napi]
 pub struct ScanResult {
+  // Private information necessary for incremental rebuilds. Note: these fields are not exposed
+  // to JS
+  base: String,
+  sources: Vec<GlobEntry>,
```
Mind = blown

Didn't know this was possible, but it keeps the information private. It's not exposed to node land.

RobinMalfait commented on crates/node/src/lib.rs:
```diff
@@ -32,35 +38,37 @@ pub struct ScanResult {
 pub struct GlobEntry {
   pub base: String,
-  pub glob: String,
+  pub pattern: String,
```
We had to do some renames already, so made this changes as well.

RobinMalfait commented on crates/oxide/src/lib.rs:
```diff
@@ -441,26 +443,6 @@ pub fn scan_files(input: Vec<ChangedContent>, options: u8) -> Vec<String> {
 }
 
-#[tracing::instrument(skip(input, globs))]
-pub fn scan_files_with_globs(input: Vec<ChangedContent>, globs: Vec<GlobEntry>) -> Vec<String> {
```
Not used right now, we can add it back when we need it.

RobinMalfait committed:
update changelog

philipp-spiess commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -228,23 +228,13 @@ export async function handle(args: Result<ReturnType<typeof options>>) {
-                glob,
-              })),
-            })
+            let candidates = scanDirResult.scanFiles(changedFiles)
```
Nice! This properly feels like "something we need to fix on the rust side" now :)

philipp-spiess commented on packages/@tailwindcss-vite/src/index.ts:
```diff
@@ -87,28 +87,28 @@ export default function tailwindcss(): Plugin[] {
-      base: inputBasePath, // Root directory, mainly used for auto content detection
-      contentPaths: globs.map((glob) => ({
+      // detection.
+      base: config!.root, // Root directory, mainly used for auto content detection
```
Maybe we can for now just disable auto-content when base is set to an empty string? We only need that API for this caller anyways.

RobinMalfait replied:
Good idea, instead of an empty string I made it an `Option<…>` in Rust and just removed it from here.

This will now disable auto content detection.

philipp-spiess reviewed (approved):
![Happy Patton Oswalt](https://media3.giphy.com/media/oirLISmToyoeI/giphy.gif?cid=47028fa8sx12d824xrvy8zo8bq3w00f23hoj11gqqzus7q0f&ep=v1_gifs&rid=giphy.gif&ct=g)

RobinMalfait committed:
make `base` path optional

When you don't provide a `base` path, we disable auto content detection.
This is currently only done in the `@tailwindcss/vite` package.

Note: you can't enable/disable auto content detection from user land
yet. That will be a separate PR.

RobinMalfait committed:
actually scan the `ChangedContent`

and later merge the candidates correctly by keeping them unique.

RobinMalfait committed:
collect changed content in vite plugin

We only have to do this if the `scanDirResult` is not ready yet. This
can happen if we `scan()` files _before_ we trigger a `generateCss` call.

If we do have `scanDirResult` enabled, then we can scan the files
directly instead of collecting them first.
